zcash_primitives/src/sapling/keys.rs

```diff
@@ -254,6 +257,72 @@ impl Diversifier {
     }
 }
 
+/// $\mathsf{KA}^\mathsf{Sapling}.\mathsf{SharedSecret} := \mathbb{J}^{(r)}$
+///
+/// Defined in [section 5.4.5.3: Sapling Key Agreement][concretesaplingkeyagreement].
+///
+/// [concretesaplingkeyagreement]: https://zips.z.cash/protocol/protocol.pdf#concretesaplingkeyagreement
+#[derive(Debug)]
+pub struct SharedSecret(jubjub::SubgroupPoint);
+
+impl SharedSecret {
+    /// TODO: Remove once public API is changed.
+    pub(crate) fn from_inner(epk: jubjub::SubgroupPoint) -> Self {
+        SharedSecret(epk)
+    }
+
+    /// For checking test vectors only.
+    #[cfg(test)]
+    pub(crate) fn to_bytes(&self) -> [u8; 32] {
+        self.0.to_bytes()
+    }
+
+    /// Only for use in batched note encryption.
+    pub(crate) fn batch_to_affine(
+        shared_secrets: Vec<Option<Self>>,
+    ) -> impl Iterator<Item = Option<jubjub::AffinePoint>> {
+        // Filter out the positions for which ephemeral_key was not a valid encoding.
+        let secrets: Vec<_> = shared_secrets
+            .iter()
+            .filter_map(|s| s.as_ref().map(|s| jubjub::ExtendedPoint::from(s.0)))
+            .collect();
+
+        // Batch-normalize the shared secrets.
+        let mut secrets_affine = vec![jubjub::AffinePoint::identity(); secrets.len()];
+        group::Curve::batch_normalize(&secrets, &mut secrets_affine);
+
+        // Re-insert the invalid ephemeral_key positions.
+        let mut secrets_affine = secrets_affine.into_iter();
+        shared_secrets
+            .into_iter()
+            .map(move |s| s.and_then(|_| secrets_affine.next()))
+    }
+
+    /// Defined in [Zcash Protocol Spec § 5.4.5.4: Sapling Key Agreement][concretesaplingkdf].
+    ///
+    /// [concretesaplingkdf]: https://zips.z.cash/protocol/protocol.pdf#concretesaplingkdf
+    pub(crate) fn kdf_sapling(self, ephemeral_key: &EphemeralKeyBytes) -> Blake2bHash {
+        Self::kdf_sapling_inner(
+            jubjub::ExtendedPoint::from(self.0).to_affine(),
```

Comment:
Is this a type-safety/spec-accuracy improvement relative to https://github.com/zcash/librustzcash/pull/753/commits/7fb80d55d64c3d25ff73542cc671c1f224372584#diff-9674216338af68a74b0dc7f5ce9b3f070551ca9971e2a5334a45cb5a6beb9127L89? I can't tell whether this is the same.

Reply:
[jubjub::SubgroupPoint::to_bytes](https://github.com/zkcrypto/jubjub/blob/6af5abf31c38e2ee006314743941c2912e39cfc4/src/lib.rs#L1435-L1437), which delegates to [jubjub::ExtendedPoint::to_bytes](https://github.com/zkcrypto/jubjub/blob/6af5abf31c38e2ee006314743941c2912e39cfc4/src/lib.rs#L1419-L1421), has to do an extended-to-affine conversion, and that [involves expensive field inversions](https://github.com/zkcrypto/jubjub/blob/6af5abf31c38e2ee006314743941c2912e39cfc4/src/lib.rs#L227-L243). In contrast [jubjub::AffinePoint::to_bytes](https://github.com/zkcrypto/jubjub/blob/main/src/lib.rs#L455-L464) does not involve any field operations. So for efficiency we want to have `kdf_sapling_inner`, defined below, for the case where we already have an `AffinePoint` (e.g. obtained by *batch* extended-to-affine conversion).

We want to avoid duplicating the body of `kdf_sapling_inner`, but we cannot use `AffinePoint::from(SubgroupPoint)` directly because there is no `impl From<SubgroupPoint> for AffinePoint`, and we cannot access `SubgroupPoint`'s `.0` member directly. Hence this indirection. I've checked it is correct and is doing the same thing as the existing code.